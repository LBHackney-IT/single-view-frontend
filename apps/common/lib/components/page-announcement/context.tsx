import React, {
  Dispatch,
  FC,
  Reducer,
  createContext,
  useCallback,
  useContext,
  useMemo,
  useReducer,
} from "react";

export interface PageAnnouncementState {
  heading: string;
  description?: string;
  variant?: "success" | "warning" | "info";
}

export type PageAnnouncementActions =
  | {
      type: "ADD";
      payload: PageAnnouncementState;
    }
  | { type: "CLEAR" };

export interface PageAnnouncementContextState {
  state?: PageAnnouncementState;
  dispatch: Dispatch<PageAnnouncementActions>;
}

export const PageAnnouncementContext = createContext<
  PageAnnouncementContextState | undefined
>(undefined);
PageAnnouncementContext.displayName = "PageAnnouncementContext";

interface UsePageAnnouncementValue {
  state?: PageAnnouncementState;
  addAnnouncement: Dispatch<PageAnnouncementState>;
  clearAnnouncement: Dispatch<void>;
}

export const usePageAnnouncement = (): UsePageAnnouncementValue => {
  const context = useContext(PageAnnouncementContext);

  if (!context) {
    const error = new Error(
      "usePageAnnouncementContext: `context` is undefined. Seems you forgot to wrap component within the Provider",
    );
    Error.captureStackTrace?.(error, usePageAnnouncement);
    throw error;
  }

  const { state, dispatch } = context;

  const addAnnouncement = useCallback(
    (props: PageAnnouncementState) => {
      dispatch({ type: "ADD", payload: props });
    },
    [dispatch],
  );

  const clearAnnouncement = useCallback(() => {
    dispatch({ type: "CLEAR" });
  }, [dispatch]);

  return {
    state,
    addAnnouncement,
    clearAnnouncement,
  };
};

interface PageAnnouncementProviderProps {
  sessionKey?: string;
}

export const PageAnnouncementProvider: FC<PageAnnouncementProviderProps> = ({
  sessionKey,
  children,
}): JSX.Element => {
  const reducer: Reducer<PageAnnouncementState | undefined, PageAnnouncementActions> = (
    state,
    action,
  ) => {
    switch (action.type) {
      case "ADD":
        return action.payload;
      case "CLEAR":
      default:
        return undefined;
    }
  };

  const initialData = useMemo(() => {
    if (sessionKey) {
      const data = {
        heading: window.sessionStorage.getItem(`${sessionKey}:heading`) || "",
        variant:
          (window.sessionStorage.getItem(
            `${sessionKey}:variant`,
          ) as PageAnnouncementState["variant"]) || "success",
        description: window.sessionStorage.getItem(`${sessionKey}:description`) || "",
      };

      if (data.heading) {
        window.sessionStorage.removeItem(`${sessionKey}:heading`);
        window.sessionStorage.removeItem(`${sessionKey}:variant`);
        window.sessionStorage.removeItem(`${sessionKey}:description`);
        return data;
      }
    }
    return undefined;
  }, [sessionKey]);

  const [state, dispatch] = useReducer(reducer, initialData);
  const value = useMemo(() => ({ state, dispatch }), [state, dispatch]);
  return (
    <PageAnnouncementContext.Provider value={value}>
      {children}
    </PageAnnouncementContext.Provider>
  );
};
