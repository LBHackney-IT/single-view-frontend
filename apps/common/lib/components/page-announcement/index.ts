export * from "./context";
export * from "./page-announcement";
