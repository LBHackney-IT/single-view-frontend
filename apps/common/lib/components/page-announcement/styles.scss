@import "~lbh-frontend/lbh/base";
@import "~lbh-frontend/lbh/components/lbh-page-announcement/page-announcement.scss";

.lbh-dialog__close {
  margin-top: 1.5rem;
  margin-right: 1.5rem;
}

.lbh-dialog__title {
  margin-right: 2rem;
}
