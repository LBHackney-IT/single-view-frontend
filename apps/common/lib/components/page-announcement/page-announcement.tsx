import React, { ComponentPropsWithoutRef, forwardRef, useContext } from "react";

import cn from "classnames";

import { PageAnnouncementContext, PageAnnouncementState } from "./context";
import "./styles.scss";

interface PageAnnouncementProps
  extends Partial<PageAnnouncementState>,
    ComponentPropsWithoutRef<"section"> {}

export const PageAnnouncement = forwardRef<HTMLElement, PageAnnouncementProps>(
  function PageAnnouncement({ className, ...props }, ref) {
    const context = useContext(PageAnnouncementContext);

    if (!context?.state?.heading && !props.heading) {
      return null;
    }

    const {
      heading,
      description,
      variant = "success",
      ...rest
    } = {
      ...context?.state,
      ...props,
    };

    return (
      <section
        ref={ref}
        className={cn(
          "lbh-page-announcement",
          {
            [`lbh-page-announcement--${variant}`]: variant && variant !== "success",
          },
          className,
        )}
        {...rest}
        role="alert"
      >
        <h3 className="lbh-page-announcement__title">{heading}</h3>
        {!!description && (
          <div className="lbh-page-announcement__content">{description}</div>
        )}
      </section>
    );
  },
);
