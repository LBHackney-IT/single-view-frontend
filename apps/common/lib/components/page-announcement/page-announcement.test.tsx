import React from "react";

import { render, testA11y } from "@hackney/mtfh-test-utils";
import { screen, waitFor } from "@testing-library/react";
import { act, renderHook } from "@testing-library/react-hooks";

import { PageAnnouncementProvider, usePageAnnouncement } from "./context";
import { PageAnnouncement } from "./page-announcement";

test("it renders correctly", async () => {
  const { container } = render(
    <PageAnnouncement heading="Announcement" description="A description" />,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it applies the correct variant", () => {
  const { container } = render(
    <PageAnnouncement heading="Announcement" variant="warning" />,
  );
  expect(container).toMatchSnapshot();
});

test("it does not render if no context or heading", () => {
  const { container } = render(<PageAnnouncement variant="warning" />);
  expect(container).toMatchSnapshot();
});

test("it can initialise with session storage", async () => {
  window.sessionStorage.setItem("test:heading", "Session Heading");
  window.sessionStorage.setItem("test:description", "Session Description");
  window.sessionStorage.setItem("test:variant", "info");
  render(
    <PageAnnouncementProvider sessionKey="test">
      <PageAnnouncement />
    </PageAnnouncementProvider>,
  );
  await waitFor(() =>
    expect(screen.getByRole("heading")).toHaveTextContent("Session Heading"),
  );
  expect(window.sessionStorage.getItem("test:heading")).toBe(null);
});

test("it wont render if session storage is empty", () => {
  const { container } = render(
    <PageAnnouncementProvider sessionKey="test">
      <PageAnnouncement />
    </PageAnnouncementProvider>,
  );
  expect(container).toMatchSnapshot();
});

test("it can set the context with usePageAnnouncement", () => {
  const wrapper: React.FC = ({ children }) => (
    <PageAnnouncementProvider>{children}</PageAnnouncementProvider>
  );
  const { result } = renderHook(() => usePageAnnouncement(), { wrapper });

  act(() => {
    result.current.addAnnouncement({
      heading: "Hook Heading",
      variant: "info",
    });
  });

  expect(result.current.state).toStrictEqual({
    heading: "Hook Heading",
    variant: "info",
  });

  act(() => {
    result.current.clearAnnouncement();
  });

  expect(result.current.state).toBe(undefined);
});

test("it throws an error when using usePageAnnouncement outside of the provider", () => {
  const { result } = renderHook(() => usePageAnnouncement());
  expect(result.error).toEqual(
    Error(
      "usePageAnnouncementContext: `context` is undefined. Seems you forgot to wrap component within the Provider",
    ),
  );
});
