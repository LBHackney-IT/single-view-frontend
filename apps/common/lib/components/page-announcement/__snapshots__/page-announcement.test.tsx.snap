// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it applies the correct variant 1`] = `
<div>
  <section
    class="lbh-page-announcement lbh-page-announcement--warning"
    role="alert"
  >
    <h3
      class="lbh-page-announcement__title"
    >
      Announcement
    </h3>
  </section>
</div>
`;

exports[`it does not render if no context or heading 1`] = `<div />`;

exports[`it renders correctly 1`] = `
<div>
  <section
    class="lbh-page-announcement"
    role="alert"
  >
    <h3
      class="lbh-page-announcement__title"
    >
      Announcement
    </h3>
    <div
      class="lbh-page-announcement__content"
    >
      A description
    </div>
  </section>
</div>
`;

exports[`it wont render if session storage is empty 1`] = `<div />`;
